core/Pos/Core/Genesis/Types.hs

```diff
@@ -155,37 +161,153 @@ bootDustThreshold (GenesisWStakeholders bootHolders) =
+    -- ^ Genesis 'BlockVersionData'.
+    , gsProtocolConstants :: !ProtocolConstants
+    -- ^ Other constants which affect consensus.
+    , gsInitializer       :: !GenesisInitializer
+    -- ^ Other data which depend on genesis type.
+    } deriving (Show, Generic)
+
+-- | Safe constructor for 'GenesisSpec'. Throws error if something
 -- goes wrong.
-mkGenesisCoreData ::
-       [AddrDistribution]
-    -> Map StakeholderId Word16
+mkGenesisSpec
+    :: GenesisAvvmBalances
+    -> SharedSeed
     -> GenesisDelegation
-    -> Either String GenesisCoreData
-mkGenesisCoreData distribution bootStakeholders delega = do
-    for_ distribution $ \(addrs, distr) -> do
-        -- Every set of addresses should match the stakeholders count
-        unless (fromIntegral (length addrs) == getDistributionSize distr) $
-            Left "mkGenesisCoreData: addressCount != stakeholdersCount \
-                 \for some set of addresses"
-        -- Addresses in each list are distinct (except for CustomBalances)
-        let isCustom = case distr of
-                CustomBalances{} -> True
-                _                -> False
-        unless (isCustom || allDistinct addrs) $
-            Left "mkGenesisCoreData: addresses in some list aren't distinct"
-        unless (all isBootstrapEraDistrAddress addrs) $
-            Left $ "mkGenesisCoreData: there is an address with stake " <>
-                   "distribution different from BootstrapEraDistr"
-
-    -- No address belongs to more than one distribution
-    let addrList = concatMap (ordNub . fst) distribution
-    unless (allDistinct addrList) $
-        Left "mkGenesisCoreData: some address belongs to more than one distr"
+    -> BlockVersionData
+    -> ProtocolConstants
+    -> GenesisInitializer
+    -> Either String GenesisSpec
+mkGenesisSpec avvmDistr seed delega bvd pc specType = do
+    let avvmKeys = HM.keys $ getGenesisAvvmBalances avvmDistr
+    unless (allDistinct avvmKeys) $
+        throwError $ "mkGenesisSpec: there are duplicates in avvm balances"
+
     -- All checks passed
-    pure
-        UnsafeGenesisCoreData
-        { gcdAddrDistribution = distribution
-        , gcdBootstrapStakeholders = GenesisWStakeholders bootStakeholders
-        , gcdHeavyDelegation = delega
-        }
+    pure $ UnsafeGenesisSpec avvmDistr seed delega bvd pc specType
+
+----------------------------------------------------------------------------
+-- GenesisData
+----------------------------------------------------------------------------
+
+-- | Genesis data contains all data which determines consensus
+-- rules. It must be same for all nodes. It's used to initialize
+-- global state, slotting, etc.
+data GenesisData = GenesisData
+    { gdBootStakeholders :: !GenesisWStakeholders
+    , gdHeavyDelegation  :: !GenesisDelegation
+    , gdStartTime        :: !Timestamp
+    , gdVssCerts         :: !VssCertificatesMap
+    , gdNonAvvmBalances  :: !(HashMap Address Coin)
```

Comment:
This seems to be the only thing that `GenesisData` has which `GenesisSpec` with a `MainnetInitializer` does not have. Should such a `HashMap Address Coin` also go in `MainnetInitializer`?

Reply:
No, in mainnet only AVVM users have coins.